Port to 1.21.1

IThundxr opened the pull request:
1.21 Neoforge is currently not out, once it is i will re-enable that subproject and finish porting

IThundxr committed:
Port to 1.20.4

IThundxr committed:
Obfuscated Obfuscation

- Fix ObfuscationReflectionHelper#setPrivateValue using SRG mapped names, Neoforge has switched to mojmap at runtime

IThundxr committed:
fixme fixedme fixed...

IThundxr committed:
port rebase fixes

IThundxr committed:
fix deps

IThundxr committed:
fix issues here and there

IThundxr committed:
fix param name

IThundxr committed:
optimize imports

IThundxr committed:
fix mixin

IThundxr committed:
fix param names

IThundxr committed:
fix instance textures

IThundxr committed:
Merge branch 'refs/heads/1.20/dev' into feat/multi-loader-1.20.4

# Conflicts:
#	gradle.properties

IThundxr committed:
build: update dependencies

IThundxr committed:
1.21 Port I

IThundxr committed:
1.21 Port II

IThundxr committed:
1.21 Port III

IThundxr committed:
1.21 Port IV

IThundxr committed:
1.21 Port V

IThundxr committed:
Mixin Fixin'

IThundxr committed:
style: javadoc fixes

IThundxr committed:
string substitution fixes

IThundxr committed:
feat(api): Make PartialModel's use RL's instead of MRL's

IThundxr committed:
Merge remote-tracking branch 'refs/remotes/upstream/1.20/dev' into feat/multi-loader-1.21

IThundxr committed:
Merge remote-tracking branch 'refs/remotes/upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	common/src/lib/java/dev/engine_room/flywheel/lib/material/Materials.java
#	common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshHelper.java
#	common/src/lib/java/dev/engine_room/flywheel/lib/visual/component/ShadowComponent.java
#	common/src/main/java/dev/engine_room/flywheel/impl/mixin/LevelRendererMixin.java
#	common/src/main/java/dev/engine_room/flywheel/impl/mixin/visualmanage/SectionCompilerMixin.java
#	forge/src/api/java/dev/engine_room/flywheel/api/event/BeginFrameEvent.java
#	forge/src/api/java/dev/engine_room/flywheel/api/event/RenderStageEvent.java
#	forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/PartialModelEventHandler.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlwImplXplatImpl.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/mixin/sodium/ChunkBuilderMeshingTaskMixin.java

IThundxr committed:
merge: post merge fixes & porting

IThundxr committed:
Merge remote-tracking branch 'refs/remotes/upstream/1.20/dev' into feat/multi-loader-1.21

IThundxr committed:
fix(mixins): broken LevelRenderer injector

IThundxr committed:
fix view -IMS

IThundxr committed:
uncomment the comments

"i commented out the particle stuff for testing something
 uh you can revert that" -IMS

IThundxr committed:
update neoforge & fix mixins

IThundxr committed:
Fix lighting

IThundxr committed:
fix jittery visual animations

IThundxr committed:
Sodium 0.6

IThundxr committed:
iris support

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/ForgeFlwConfig.java

IThundxr committed:
forge porting

jdnjk commented:
forge?

IThundxr commented:
> forge?

There will be no support for MCForge, only fabric and neoforge will be supported

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	buildSrc/build.gradle.kts
#	buildSrc/src/main/kotlin/dev/engine_room/gradle/subproject/SubprojectPlugin.kt
#	forge/build.gradle.kts
#	gradle.properties
#	gradle/wrapper/gradle-wrapper.properties

IThundxr committed:
Add sodium 0.6 compat

IThundxr committed:
Fix issues

IThundxr committed:
fix remapping

IThundxr committed:
Merge branch '1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	common/src/lib/java/dev/engine_room/flywheel/lib/model/part/ModelPartConverter.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/ForgeFlwConfig.java
#	gradle.properties

IThundxr committed:
Fix compile errors

IThundxr committed:
Merge branch '1.20/dev' into feat/multi-loader-1.21

IThundxr committed:
port changes to 1.21

IThundxr committed:
Merge branch '1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	forge/src/main/java/dev/engine_room/flywheel/impl/ForgeFlwConfig.java

IThundxr committed:
post merge fixes

IThundxr committed:
Fix sprite UVs being messed up

IThundxr committed:
Fix iris compat

IThundxr committed:
Build against 1.21.1

IThundxr committed:
Last fixes

Jozufozu commented on forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java:
```diff
@@ -15,45 +17,43 @@
 
-	public FlywheelForge() {
+	public FlywheelForge(IEventBus modEventBus, ModContainer modContainer) {
 		ModLoadingContext modLoadingContext = ModLoadingContext.get();
```
nit: unused

Jozufozu commented on forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java:
```diff
@@ -15,45 +17,43 @@
+import net.neoforged.neoforge.registries.RegisterEvent;
 
 @Mod(Flywheel.ID)
 public final class FlywheelForge {
```
We should rename all the Forge* classes to be NeoForge*

Jozufozu commented on common/src/main/java/dev/engine_room/flywheel/impl/compat/CompatMods.java:
```diff
@@ -0,0 +1,22 @@
+	private final Supplier<Boolean> isLoaded;
+
+	CompatMods() {
+		isLoaded = FlwImplXplat.INSTANCE.getModLoaded(name().toLowerCase(Locale.ROOT));
```
I feel like it's more sound to include the proper mod ID as an argument to the enum

Jozufozu commented on common/src/main/java/dev/engine_room/flywheel/impl/compat/CompatMods.java:
```diff
@@ -0,0 +1,22 @@
+	EMBEDDIUM,
+	IRIS;
+
+	private final Supplier<Boolean> isLoaded;
```
nit: BooleanSupplier

Jozufozu commented on forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java:
```diff
@@ -68,10 +68,10 @@ private static void clientInit(IEventBus forgeEventBus, IEventBus modEventBus) {
 			// Make sure we don't tick on the server somehow.
-			if (e.phase == TickEvent.Phase.END && e.side == LogicalSide.CLIENT) {
-				VisualizationEventHandler.onClientTick(Minecraft.getInstance(), e.level);
+			if (FMLLoader.getDist().isClient()) {
```
Should this be
e.getLevel() instanceof ClientLevel

LevelTickEvent says it's dispatched for both the logical client and logical server so this would be fired twice in single player

VisualizationManagerImpl#get checks for that too so it works fine now be we can early out sooner

Jozufozu commented on forge/src/main/resources/META-INF/neoforge.mods.toml:
```diff
@@ -14,22 +14,24 @@ authors = "Jozufozu, PepperCode1"
-mandatory = false
-versionRange = "[0.7.0,)"
-side = "CLIENT"
+modId = "embeddium"
```
Don't we have support for embeddium?

IThundxr replied:
Embeddium is no longer actively maintained while being very out of date compared to sodium, no reason to deal with the increased amount of issues it'll generate

Jozufozu commented on buildSrc/src/main/kotlin/dev/engine_room/gradle/subproject/SubprojectPlugin.kt:
```diff
@@ -89,21 +89,25 @@ class SubprojectPlugin: Plugin<Project> {
                     includeGroup("maven.modrinth")
                 }
             }
+            flatDir {
```
nit: remove

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/part/VertexWriter.java:
```diff
@@ -23,62 +23,60 @@ public VertexWriter() {
+	public VertexConsumer endVertexIfNeeded() {
 		if (!filledPosition || !filledTexture || !filledNormal) {
-			throw new IllegalStateException("Not filled all elements of the vertex");
+			return this;
```
Will there be new subtle bugs in dev instead of exceptions because we never throw? Does this mimic some changed vanilla behavior? If so please add a comment explaining it

IThundxr replied:
Vanilla no longer throws a exception, although it'd probably be a good idea to throw in dev, even if it differs from vanilla behaviour

Jozufozu commented on forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitter.java:
```diff
@@ -39,12 +43,12 @@ public BufferBuilder unwrap(boolean shade) {
-			bufferBuilder.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.BLOCK);
+	void prepareForGeometry(boolean shade) {
+		if (!((BufferBuilderAccessor) bufferBuilder).flywheel$getBuilding()) {
+			bufferBuilder = new BufferBuilder(byteBufferBuilder, VertexFormat.Mode.QUADS, DefaultVertexFormat.BLOCK);
```
Are BufferBuilders single use now?

IThundxr replied:
Yes

IThundxr committed:
address review

IThundxr committed:
Rename class

IThundxr committed:
post to correct bus

IThundxr committed:
fix breaks formatting

IThundxr committed:
bring back comment

IThundxr committed:
forgeEventBus -> gameEventBus

IThundxr committed:
Remove unused check

MattiDragon commented:
This PR seems to have a bug that compleatly breaks the hitbox entity component. The fire one might also be broken, but I have no easy repro.
<details>
<summary>Image of the issue</summary>

![2024-10-13_20 56 13](https://github.com/user-attachments/assets/f50d7258-00f7-4868-8c3d-6a9484000e84)

</details>

Repro steps:
1. Open up the game with flywheel installed
2. Enable either the indirect or instancing backend
3. Spawn a minecart
4. Press F3+B to enable hitboxes
-> The hitbox renders as above

The issue disappears after reloading chunks with F3+A and sometimes reappears in after rejoining the game. An alternative form of the bug can also appear after F3+A with some combination of reloads and hitbox toggles, which looks like this:
<details>
<summary>Image of the alternative form</summary>

![2024-10-13_20 58 56](https://github.com/user-attachments/assets/7de7ea0e-a737-4751-abcc-df88b3275a10)

</details>
I was unable to reliably reproduce this form.

I have no idea what's going on, but its very broken.

Jozufozu commented on fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitter.java:
```diff
@@ -37,22 +41,21 @@ public BufferBuilder getBuffer(boolean shade) {
-		if (!bufferBuilder.building()) {
-			bufferBuilder.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.BLOCK);
-		} else if (shade != currentShade) {
+		bufferBuilder = new BufferBuilder(byteBufferBuilder, VertexFormat.Mode.QUADS, DefaultVertexFormat.BLOCK);
```
I think this needs to move back into the if statement.

Placing a hopper minecart:
```
Caused by: java.lang.IllegalStateException: Not building!
	at com.mojang.blaze3d.vertex.BufferBuilder.ensureBuilding(BufferBuilder.java:76) ~[minecraft-merged-b25f1a0b75-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2.jar:?]
	at com.mojang.blaze3d.vertex.BufferBuilder.beginVertex(BufferBuilder.java:95) ~[minecraft-merged-b25f1a0b75-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2.jar:?]
	at com.mojang.blaze3d.vertex.BufferBuilder.addVertex(BufferBuilder.java:156) ~[minecraft-merged-b25f1a0b75-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2.jar:?]
	at dev.engine_room.flywheel.lib.model.baked.UniversalMeshEmitter.addVertex(UniversalMeshEmitter.java:60) ~[lib/:?]
```

Jozufozu commented on forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitter.java:
```diff
@@ -39,12 +43,11 @@ public BufferBuilder unwrap(boolean shade) {
-			bufferBuilder.begin(VertexFormat.Mode.QUADS, DefaultVertexFormat.BLOCK);
-		} else if (shade != currentShade) {
+	void prepareForGeometry(boolean shade) {
+		bufferBuilder = new BufferBuilder(byteBufferBuilder, VertexFormat.Mode.QUADS, DefaultVertexFormat.BLOCK);
```
Same here.

```
Caused by: java.lang.IllegalStateException: Not building!
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.BufferBuilder.ensureBuilding(BufferBuilder.java:76) ~[neoforge-21.1.66-minecraft-merged-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2-neoforge-21.1.66.jar%23191!/:?] {re:mixin,pl:unprotect:A,pl:runtimedistcleaner:A,re:classloading,pl:mixin:APP:flywheel.impl.mixins.json:BufferBuilderAccessor from mod flywheel,pl:mixin:A,pl:unprotect:A,pl:runtimedistcleaner:A}
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.BufferBuilder.beginVertex(BufferBuilder.java:95) ~[neoforge-21.1.66-minecraft-merged-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2-neoforge-21.1.66.jar%23191!/:?] {re:mixin,pl:unprotect:A,pl:runtimedistcleaner:A,re:classloading,pl:mixin:APP:flywheel.impl.mixins.json:BufferBuilderAccessor from mod flywheel,pl:mixin:A,pl:unprotect:A,pl:runtimedistcleaner:A}
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.BufferBuilder.addVertex(BufferBuilder.java:250) ~[neoforge-21.1.66-minecraft-merged-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2-neoforge-21.1.66.jar%23191!/:?] {re:mixin,pl:unprotect:A,pl:runtimedistcleaner:A,re:classloading,pl:mixin:APP:flywheel.impl.mixins.json:BufferBuilderAccessor from mod flywheel,pl:mixin:A,pl:unprotect:A,pl:runtimedistcleaner:A}
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.VertexConsumer.putBulkData(VertexConsumer.java:142) ~[neoforge-21.1.66-minecraft-merged-1.21.1-loom.mappings.1_21_1.layered+hash.1839147940-v2-neoforge-21.1.66.jar%23191!/:?] {re:classloading,pl:runtimedistcleaner:A}
	at TRANSFORMER/flywheel@1.0.0-beta/dev.engine_room.flywheel.lib.model.baked.MeshEmitter.putBulkData(MeshEmitter.java:84) ~[%23193!/:?] {re:classloading}
```

Jozufozu commented on fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MeshEmitter.java:
```diff
@@ -3,29 +3,33 @@
 	public void end() {
-		if (bufferBuilder.building()) {
+		if (((BufferBuilderAccessor) bufferBuilder).flywheel$getBuilding()) {
 			emit();
```
This may also need to reset the buffer builder

IThundxr committed:
fix meshemitter

IThundxr committed:
formatting changes

IThundxr committed:
address requested changes

IThundxr committed:
address requested changes

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	common/src/lib/java/dev/engine_room/flywheel/lib/util/ShadersModHandler.java
#	common/src/main/java/dev/engine_room/flywheel/impl/FlwImplXplat.java
#	common/src/main/java/dev/engine_room/flywheel/impl/compat/SodiumCompat.java
#	common/src/main/java/dev/engine_room/flywheel/impl/mixin/BlockEntityTypeMixin.java
#	fabric/build.gradle.kts
#	fabric/src/main/java/dev/engine_room/flywheel/impl/FlwImplXplatImpl.java
#	fabric/src/main/java/dev/engine_room/flywheel/impl/FlwLibXplatImpl.java
#	fabric/src/main/java/dev/engine_room/flywheel/impl/mixin/sodium/ChunkBuilderMeshingTaskMixin.java
#	fabric/src/main/java/dev/engine_room/flywheel/impl/mixin/sodium/SodiumMixinPlugin.java
#	fabric/src/main/resources/fabric.mod.json
#	forge/build.gradle.kts
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlwImplXplatImpl.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlwLibXplatImpl.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java
#	forge/src/main/resources/META-INF/neoforge.mods.toml
#	gradle.properties
#	gradle/wrapper/gradle-wrapper.properties

IThundxr committed:
post merge fixes

IThundxr committed:
Twas a snowy night full of scripting the builds

- Post merge fixes
- Remove stubs

IThundxr committed:
Stubborn stubs

- Remove Stub SourceSet

PepperCode1 committed:
VertexConsumer-related clean up

- Clean up VertexWriter to closer match BufferBuilder
- Clean up MeshEmitter on Fabric/NeoForge and remove BufferBuilderAccessor
- Remove model data and model data lookup parameters in NeoForge model builders; use IBlockGetterExtension#getModelData instead
- Fix artifact Minecraft version

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	.github/workflows/build.yml
#	forge/src/main/resources/META-INF/neoforge.mods.toml

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlwCommands.java

IThundxr committed:
Mining the errors away'

post merge fixes

PepperCode1 committed:
Fix VertexWriter

- Bump pack format in pack.mcmeta

PepperCode1 commented:
- [x] Fix test mod
- [x] Inline platform-specific model builders

IThundxr committed:
Update test mod for neoforge

IThundxr committed:
Work on fixing testmod

IThundxr committed:
Rename subproject to neoforge

IThundxr committed:
Misc renames

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

IThundxr committed:
fix sodium dep

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

IThundxr committed:
Embeddium compat

IThundxr committed:
Fix sodium compat

PepperCode1 committed:
Inline platform-specific model builders

Jozufozu commented on common/src/lib/java/dev/engine_room/flywheel/lib/model/baked/MultiBlockModelBuilder.java:
```diff
@@ -24,15 +23,11 @@ public abstract class MultiBlockModelBuilder {
 		this.positions = positions;
 	}
 
-	public static MultiBlockModelBuilder create(BlockAndTintGetter level, Iterable<BlockPos> positions) {
```
Do we trust that we won't need these factory methods in the future? It's easy to change what the contents of a function point to while changing constructors is much more delicate.

Jozufozu commented on fabric/src/main/resources/fabric.mod.json:
```diff
@@ -24,16 +24,15 @@
   "breaks": {
-    "sodium": ["<0.5.0", "~0.6.0- <0.6.0-beta.2"],
+    "sodium": "<0.6.0-beta.2",
     "embeddium": "*"
```
Remove breaks embeddium. It's not even released on 1.21 fabric

IThundxr committed:
Update fabric.mod.json

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20/dev' into feat/multi-loader-1.21

# Conflicts:
#	common/src/lib/java/dev/engine_room/flywheel/lib/internal/FlwLibXplat.java
#	common/src/lib/java/dev/engine_room/flywheel/lib/model/ModelUtil.java
#	common/src/lib/java/dev/engine_room/flywheel/lib/model/Models.java
#	fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/FabricBakedModelBuilder.java
#	fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/FabricBlockModelBuilder.java
#	fabric/src/lib/java/dev/engine_room/flywheel/lib/model/baked/FabricMultiBlockModelBuilder.java
#	fabric/src/main/java/dev/engine_room/flywheel/impl/FlwLibXplatImpl.java
#	forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/ForgeBakedModelBuilder.java
#	forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/ForgeBlockModelBuilder.java
#	forge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/ForgeMultiBlockModelBuilder.java
#	forge/src/main/java/dev/engine_room/flywheel/impl/FlywheelForge.java
#	neoforge/src/lib/java/dev/engine_room/flywheel/lib/model/baked/BakedModelBufferer.java
#	neoforge/src/main/java/dev/engine_room/flywheel/impl/FlwLibXplatImpl.java

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20.1/dev' into feat/multi-loader-1.21

# Conflicts:
#	neoforge/src/main/java/dev/engine_room/flywheel/impl/FlwCommands.java
#	neoforge/src/main/java/dev/engine_room/flywheel/impl/NeoForgeFlwConfig.java

IThundxr committed:
more renames

Jozufozu committed:
Make explicit the implicit

- Add @apiNotes explaining that flywheel/ is prepended to each shader
  ResourceLocation's path
- Document all InstanceType fields

Jozufozu committed:
No need to get spicy

- Move common vanillin stuffs to a "vanillin" sourceset
- Add vanillinForge and vanillinFabric subprojects
- Mostly do gradling from scratch
- Export remap jars from platform projects to `include` in vanillin
  projects

Jozufozu committed:
See yourself out

- Replace outgoing jars by directly consuming classes/resources from
  common
- Add vanillin logo
- Move subproject plugin logic to an extension so it can be configured
- Rename "mod_" properties to "flywheel_"

Jozufozu committed:
Scratch that

- Loom gets angry when a mod* dependency doesn't immediately exist so
  use a regular runtime dependency with the devlibs jar
- Strangely, the include dependency is just fine with the remap jar
- Move outgoing remapJar/jar configuration generation to JarTaskSet
- Don't actually need evaluationDependsOn
- Fix test mod id

Jozufozu committed:
Oh my glob

- Glob over all jars from libs when running the test mod
- Add vanillin jars to artifacts

Jozufozu committed:
String theory

- Assume group/version are behind properties in subproject.init
- Force each subproject to manually specify their base archive name
- Don't generate mod/api artifact ids, just write them out by hand
- Remove DependentProject

Jozufozu committed:
Hrm

- Remove sources and javadoc after glob copying them

Jozufozu committed:
I do declare

- Move source set bundling/common source set sharing logic into
  transitiveSourceSets and make it more flexible
- Remove commonProject field from platform extension

Jozufozu committed:
One and the same

- Add model id debug view
- Want to be able to visually check that instances are pointing to the
  same model

IThundxr committed:
Merge remote-tracking branch 'upstream/1.20.1/dev' into feat/multi-loader-1.21

IThundxr committed:
misc fixes

IThundxr committed:
more fixes

IThundxr committed:
h

Jozufozu committed:
Fix dependency type declarations

Jozufozu committed:
Fix vanillin dependency type declarations

Jozufozu reviewed (approved):
I think this is good enough to merge. Any further changes can be made in-branch
